src/generation_tools.jl

```diff
@@ -43,14 +41,80 @@ julia> idx_to_dim([(x,y) for x in 1:3, y in 1:3])
  1  2  3
 ```
 """
-function idx_to_dim(idx_arr::AbstractArray{T, N}) where {T, N} 
-    out_arr = Array{typeof(idx_arr[1][1])}(undef, size(idx_arr)..., length(idx_arr[begin]))
+function idx_to_dim(idx_arr::AbstractArray{T, N}) where {TT, NT, T<:NTuple{NT, TT}, N} 
+    newdims = ntuple((d)->mod(d, N+1)+1, N+1)
+    return permutedims(idx_to_arr_view(idx_arr), newdims)
+end
+
+"""
+    dim_to_idx(v) 
+converts an Array to an Array of Tuples by packaging the outer dimension into the inner dimension of the new array.
 
-    # loop over array
-    @inbounds for I in CartesianIndices(idx_arr)
-        for (i, ind) in enumerate(axes(out_arr, N+1))
-            out_arr[Tuple(I)..., ind] = idx_arr[I][i]
-        end
-    end
-    return out_arr
+See also: `idx_to_dim` which is tarr_to_arr
+
+Example:
+```jldoctest
+julia> NDTools.arr_to_tarr([1 3 5; 2 4 6])
+3-element Vector{Tuple{Int64, Int64}}:
+ (1, 2)
+ (3, 4)
+ (5, 6)
+```
+"""
+function dim_to_idx(v)
+    newdims = ntuple((d)->mod(d-2, ndims(v))+1, ndims(v))
+    return arr_to_idx_view(permutedims(v, newdims))
+end
+
+"""
+    idx_to_arr_view(idx_arr) 
+
+Reinterprets an N-dimensional array of NTuple (idx_arr) as an N+1 dimensional array by unrolling the (inner) tuple.
+
+See also: `arr_to_idx_view`
+
+Arguments:
++ `idx_arr`. The array of NTuple to reinterpret
+
+Example:
+```jldoctest
+julia> idx_to_arr_view([(x,y) for x in 1:3, y in 1:3])
+2×3×3 reinterpret(reshape, Int64, ::Matrix{Tuple{Int64, Int64}}) with eltype Int64:
+[:, :, 1] =
+ 1  2  3
+ 1  1  1
+
+[:, :, 2] =
+ 1  2  3
+ 2  2  2
+
+[:, :, 3] =
+ 1  2  3
+ 3  3  3
+```
+"""
+function idx_to_arr_view(idx_arr::AbstractArray{T, N}) where {TT, NT, T<:NTuple{NT, TT}, N} 
```

Comment:
type instable

Reply:
Not correct. It is type stable.
`@code_warntype idx_to_arr_view([(x,y) for x in 1:3, y in 1:3])`